crates/polars-core/src/chunked_array/logical/categorical/mod.rs

```diff
@@ -48,16 +48,39 @@ impl CategoricalChunked {
         self.logical.name()
     }
 
-    /// Get a reference to the logical array (the categories).
+    // TODO: Rename this
+    /// Get a reference to the physical array (the categories).
     pub fn logical(&self) -> &UInt32Chunked {
         &self.logical
     }
 
-    /// Get a reference to the logical array (the categories).
+    /// Get a mutable reference to the physical array (the categories).
     pub(crate) fn logical_mut(&mut self) -> &mut UInt32Chunked {
         &mut self.logical
     }
 
+    /// Convert a categorical column to its local representation.
+    pub fn to_local(&self) -> Self {
+        let rev_map = self.get_rev_map();
+        let (physical_map, categories) = match rev_map.as_ref() {
+            RevMapping::Global(m, c, _) => (m, c),
+            RevMapping::Local(_) => return self.clone(),
+        };
+
+        let local_rev_map = RevMapping::Local(categories.clone());
+        // TODO: A fast path can possibly be implemented here:
+        // if all physical map keys are equal to their values,
+        // we can skip the apply and only update the rev_map to local
+        let local_ca = self.logical().apply(|v| *physical_map.get(&v).unwrap());
```

Comment:
This is my bad. 

Now that I think about it. If the array has got null values, the masked out value might be an incorrect index. So we must use `apply_on_opt` here.

Reply:
You have to create a categorical that is out of bounds under the mask, not always easy. Can checkout this PR later.

Reply:
Updated the implementation to use `apply_on_opt` 👍